### PR TITLE
Strict Comparisons

### PR DESCRIPTION
**Pull request by mattyrob**
Sniffs and Unit Tests for Strict Comparisons so === not == and !== not
!=

Replaces ticker #243 that got scrubbed inadvertently

### CODE REVIEW

**Comment by GaryJones** [comment]
Line 44, 45, 47 (and maybe others) have incorrect indentation. Part of this will be because we've not clarified which coding standards these sniffs should be following, but let's at least try to the choice of indentation character consistent within each file :-)

File's should not close with `?>` and should have a clear line at the end.

@JDGrimes can better advise, but I think we now have a `WordPress_Sniff` class you can inherit from, which already includes the method for checking for a whitelist comment. This class was added since your first PR.

Tests seem to use `Good`, but I think existing tests use `OK`?

#reviewmerge

**Comment by JDGrimes** [comment]
> File's should not close with ?> 

30 of them do though (lack of standards again).



> I think we now have a WordPress_Sniff class you can inherit from, which already includes the method for checking for a whitelist comment.

Yes. If you extend `WordPress_Sniff` you can use its [`has_whitelist_comment()` method](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniff.php#L107). You just need to call `$this->init( $phpcsFile )` at the top of `process()`, and then call `has_whitelist_comment()` like so:

```php
$this->has_whitelist_comment( 'loose comparison', $stkPtr );
```

Note that it takes only the two arguments, and that you should omit the `okay` suffix from the comment.



> Tests seem to use Good, but I think existing tests use OK?

We break that rule 36 times. We need some standards for the standards, and some standard testing standards. :smile: 

---
@mattyrob If you don't want to tackle the changes, just say the word and I'll merge and make any modifications needed.

---
Note: There is a [`Squiz.Operators.ComparisonOperatorUsage` sniff](https://github.com/squizlabs/PHP_CodeSniffer/blob/5b63fd3fe960e2020770c845e04254e5ca2b722a/CodeSniffer/Standards/Squiz/Sniffs/Operators/ComparisonOperatorUsageSniff.php), but it is much more complex, because it disallows implicit boolean comparisons (`if ( $var )/if ( ! $var )`).

File: WordPress/Sniffs/PHP/StrictComparisonsSniff.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	 * @param int			$stackPtr  The position of the current token in the
+	 *						stack passed in $tokens.
+	 *
+	 * @return void
+	 */
+	public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr)
+	{
+		$tokens = $phpcsFile->getTokens();
+
+		if ( $tokens[$stackPtr]['code'] !== T_IS_EQUAL && $tokens[$stackPtr]['code'] !== T_IS_NOT_EQUAL) {
```
Comment (JDGrimes) [inline WordPress/Sniffs/PHP/StrictComparisonsSniff.php]: I don't think that this condition is necessary. PHPCS will already be limiting the sniff to these tokens based on `register()`, won't it?

**Comment by GaryJones** [comment]
> 30 of them do though (lack of standards again).

That's a WP and a PSR standard, and the one file in PHPCS I just checked also drops it, so there's no reason not to follow it too.

> We break that rule 36 times. We need some standards for the standards, and some standard testing standards.

Do we have an existing ticket for our own code standards? Thought we did, but it may just be comments on another ticket. Just like PHPCS has a PHPCS Standard that itself follows, we can do the same with a WPCS standard, where we can mostly use the WordPress standard we're familiar with, but allow for camel case method names and other things needed as we extend PHPCS.

**Comment by JDGrimes** [comment]
> Do we have an existing ticket for our own code standards?

#256

**Comment by mattyrob** [comment]
I've made some changes based on the suggestions above. Feel free to make further ones as you need.

**Comment by GaryJones** [comment]
The tests are currently failing, as you've changed "loose comparison okay" whitelisting comment to "loose comparison, OK", without using the `WordPress_Sniff` defined class to define the whitelisting comment method.

**Comment by mattyrob** [comment]
Fix for the synced

**Comment by JDGrimes** [comment]
Great, I'll take it from there.

**Comment by mattyrob** [comment]
@JDGrimes - thanks for that ;)

**Comment by westonruter** [comment]
> The tests are currently failing, as you've changed "loose comparison okay" whitelisting comment to "loose comparison, OK", without using the WordPress_Sniff defined class to define the whitelisting comment method.

I don't think we need a whitelisting comment. The ambiguity should be resolved with code, for instance if `$value` is a string `'1'` then equality should be checked via:

```
if ( 1 === intval( $value ) ) {
    // ...
}
```

This eliminates all ambiguity.

**Pull request merged**